src/librustdoc/html/highlight.rs

```diff
@@ -748,11 +796,51 @@ fn string<T: Display>(
                 }
             })
         {
-            write!(out, "<a class=\"{}\" href=\"{}\">{}</a>", klass.as_html(), href, text_s);
-            return;
+            write!(out, "<a class=\"{}\" href=\"{}\">{}", klass.as_html(), href, text_s);
+            return Some("</a>");
         }
     }
-    write!(out, "<span class=\"{}\">{}</span>", klass.as_html(), text_s);
+    write!(out, "<span class=\"{}\">{}", klass.as_html(), text_s);
+    Some("</span>")
+}
+
+/// This function is to get the external macro path because they are not in the cache used n
+/// `href_with_root_path`.
+fn generate_macro_def_id_path(context_info: &ContextInfo<'_, '_, '_>, def_id: DefId) -> String {
```

Comment:
I'm not super happy about this part: the `cache` contains external items but not macros, meaning that I had to compute it here. I originally thought about adding the macro paths into the `cache` but after a discussion with @jyn514, we thought that maybe it would be worth it to compute macros' paths on demand.

To be noted: this is a mix between `clean::inline::record_extern_fqn` and `html::format::href_with_root_path`. Does it look reasonable as is?